[clangd] Autocomplete fixes for methods

etyloppihacilem opened the pull request:
# Clangd completion fixes

Various fixes in clangd completion, including some changes to the Sema lib.

## fix clangd/clangd#1752

Adds support to autocomplete method and function arguments and qualifiers.

Edited unittests to match new behavior.

Added `CodeCompletionString::CK_FunctionQualifier` to differentiate between plain `CodeCompletionString::CK_Informative` (e.g. `", ..."`) and function qualifiers.
No other change to Sema behavior.

## fix clangd/clangd#880

Change completion context if a method definition is detected, to allow private/protected completion.

Change to context is only temporary. No other impact on Sema lib behavior.

## fix clangd/clangd#753

Complete arguments with default value in case of definition, removing the values.

Added unittests.

```cpp
// example.hpp
class B {
    public:
        B(int a, int b);
};

class Test
{
public:
    void foo(int a, int b, int c);
    void bar(int a, int b, int c) const;
    void defargs(int a, int b, int c = 0, int d=1);
    void defargs2(int a, int b = 1, B obj = B(1, 2), int c = 5);
private:
    void priv(int a, int b);
};
```

```cpp
// example.cpp

void Test::foo^
// completes
void Test::foo(int a, int b, int c)

void Test::bar^
// completes
void Test::bar(int a, int b, int c) const

void Test::defargs^
// completes
void Test::defargs(int a, int b, int c, int d)

void Test::defargs2^
// completes
void Test::defargs2(int a, int b, B obj, int c)

void Test::priv^
// completes
void Test::priv(int a, int b)
```

It does respect the `--function-arg-placeholders` argument and the `ArgumentLists` parameter.

I don't know if those are unrelated changes, should I maybe open different PR for those three issues ?

etyloppihacilem committed:
[clangd] Autocomplete fixes for methods and arguments

etyloppihacilem commented:
Ping

HighCommander4 reviewed (changes requested):
Thanks for working on this, it's an important issue that has been bothering a lot of people!

Unfortunately, the current implementation approach regresses the current (intentional) behaviour that when you're taking the address of a function, function parameters are **not** inserted.

For example:

```c++
struct Test {
    void fun(int a, int b, int c);
};

int main() {
  Test t;
  &Test::f^
}
```

The expected completion here is just `fun`, but with this patch it will be `fun(int a, int b, int c)`.

(This is reflected in the test case `CompletionTest.HeuristicsForMemberFunctionCompletion`. We want to keep this test case passing without changes to the expectations.)

I think the way this should be implemented is that in SemaCodeComplete, we determine whether we are completing the declaration/definition of a function, or a use of it.

Sam gave a hint for how to do this in [this comment](https://github.com/clangd/clangd/issues/880#issuecomment-961533830):

> It looks like ParseOptionalCXXScopeSpecifier also doesn't know, so maybe the right thing is to add yet another boolean parameter there and pass `true` if we're parsing a declarator.

(`ParseOptionalCXXScopeSpecifier` is the caller of `SemaCodeCompletion::CodeCompleteQualifiedId`, so the idea would be to pass this new boolean parameter into `CodeCompleteQualifiedId`.)

Once we have this piece of knowledge available in SemaCodeComplete, we can implement more of the "definition vs. use" logic there. For example, rather than doctoring the code completion string in clangd to remove the placehoders in the "definition" case, we can just not add placeholder chunks in the first place.

If you'd like to give implementing this approach a try and have some follow-up questions, please don't hesitate to ask, either here or in the clangd Discord channel.

etyloppihacilem commented:
Just corrected the unused variable warning, feel free to ping me on discord `@etyloppihacilem`

HighCommander4 commented on clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp:
```diff
@@ -585,39 +591,94 @@ TEST(CompletionTest, HeuristicsForMemberFunctionCompletion) {
+    EXPECT_THAT(
+        Results.Completions,
+        Contains(AllOf(named("staticGeneric"),
+                       signature("<typename T, int U>()"), snippetSuffix(""))));
```
Similarly, I would still expect `staticGeneric` to have a snippet for the template argument list in a no-call context.

HighCommander4 reviewed (commented):
Thanks for the update!

I haven't had a chance to look at the implementation yet (I hope to do so over the next few days), I just looked at the changes to CodeCompleteTests.cpp, and wrote down a few questions I had so far.

etyloppihacilem commented:
Thank you for your time and your detailed replies, I effectively mixed things up with templates and I'm sorry about that. I will try to fix the template situation today or in the days to come.

I figured another problematic case :

```cpp
template < typename T >
class Test {
    public:
        Test();
};

template < typename T >
Tes^
```
Where it wrongly completes
```cpp
Test<typename T> // with placeholder around template argument
```
It should look something like
```cpp
Test<T>
```

etyloppihacilem commented:
I fixed the typo and every template related issue, except the last one I pointed out.

HighCommander4 commented on clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp:
```diff
@@ -585,39 +591,93 @@ TEST(CompletionTest, HeuristicsForMemberFunctionCompletion) {
-                Contains(AllOf(named("method"), signature("(int) const"),
+                Contains(AllOf(named("method"), signature("(int name) const"),
                                snippetSuffix(""))));
-    // We don't have any arguments to deduce against if this isn't a call.
```
Please retain this comment

HighCommander4 commented on clang-tools-extra/clangd/unittests/CodeCompletionStringsTests.cpp:
```diff
@@ -158,6 +160,28 @@ TEST_F(CompletionStringTest, SnippetsInPatterns) {
+      *Builder.TakeString(),
+      /*ResultKind=*/CodeCompletionResult::ResultKind::RK_Declaration,
+      /*IncludeFunctionArguments=*/false, /*IsDefinition=*/true);
+  // Arguments placeholders dropped from snippet, kept in signature.
```
The comment here should say "dropped from snippet and signature"

HighCommander4 commented:
I've started to look at the implementation. I haven't looked through everything in detail yet, but the main thing I'm noticing so far is that the patch doesn't actually implement Sam's suggestion of having the parser pass in to `CodeCompleteQualifiedId` a flag indicating that we are in a declaration context.

Instead, the patch is inferring this information inside `CodeCompleteQualifiedId` by doing this:

```c++
  if (!SemaRef.CurContext->isFunctionOrMethod() &&
      !SemaRef.CurContext->isRecord()) {
    // Assume we are completing a declaration
  }
```

But this has false positives. For example:

```c++
namespace N {
  int foo(int required, int optional = 42);
}

int foo = N::f^
```

Here, the behaviour before the patch was to produce a `DeclaringEntity=false` completion, giving us just the required argument and making it a placeholder.

The behaviour after the patch is to produce a `DeclaringEntity=true` completion, inserting both arguments with no placeholders.

I think it would be better to keep this a `DeclaringEntity=false` completion by using a more accurate determination of whether we are in a declaration context, by having the parser pass in a boolean to `CodeCompleteQualifiedId` indicating this. (Note, this would be an additional boolean to `IsAddressOfOperand`, something like `IsInDeclarationContext`.)

etyloppihacilem commented:
I decided to use `Parser::ParseDeclaratorInternal` as a criteria to detect declaration context, it works great and was way easier than my previous approach.


I had to fix another issue where static and non-static member functions were mixed in the completion results. I now detect if a call occurs within a scope resolution context (::) to treat non-static functions like pointers, i.e., no completion (matches the behavior before my patch). This distinction is clearly visible here:
```cpp
struct Foo {
  template <typename T, typename U, typename V = int>
  T generic(U nameU, V nameV);
};

int i = Foo::^
// would complete
int i = Foo::generic<typename T, typename U>(U nameU, V nameV) // with placeholders
// which is incorrect because in a call context, we do not want 'typename U'
// and the call context is incorrect anyway

// it now complete, as it did before
int i = Foo::generic

// while
int main() {
  Foo f;
  f.^
  // completes
  f.generic<typename T>(U nameU, V nameV) // with placeholders as well.
}
```

HighCommander4 commented:
Thanks!

One question: is the new flag `CodeCompletionResult::IsAddressOfOperand` giving us anything that setting `DeclaringEntity=false` and `FunctionCanBeCall=false` wouldn't?

My other high-level piece of feedback is that I think we could implement some of this functionality more cleanly at the time a `CodeCompletionString` is constructed. Specifically, in [`CodeCompletionResult::createCodeCompletionStringForDecl`](https://searchfox.org/llvm/rev/b397c9d24196fe4103ad7cfe6bbfdfc08496364b/clang/lib/Sema/SemaCodeComplete.cpp#3748), we have access to flags such as `DeclaringEntity` and `FunctionCanBeCall` (they are fields of `this`), and we can pass them in to helpers such as `AddFunctionParameterChunks` and `AddFunctionTypeQualsToCompletionString` and let them influence the string chunks directly. For example:

 * If `DeclaringEntity=true`, instead of adding `CK_Placeholder` chunks for function parameters, we can create `CK_Text` chunks.
 * For parameters with default values, instead of adding the parameter and default value as a single chunk, we can add them as two separate chunks, with the default value being `CK_Informative`.
    * Clangd puts `CK_Informative` chunks into [the signature but not the snippet](https://searchfox.org/llvm/rev/b397c9d24196fe4103ad7cfe6bbfdfc08496364b/clang-tools-extra/clangd/CodeCompletionStrings.cpp#277-283), which is what we want.
 * For the function qualifiers, we can add them as `CK_Text` if `DeclaringEntity=true`, and otherwise as `CK_Informative`. (We don't need to introduce the new `CK_FunctionQualifier` chunk kind.)

With this approach, most (all?) changes to the clangd `getSignature` function should become unnecessary, and other consumers of the Sema code completion API (e.g. libclang) get to benefit from our improvements as well.

WDYT?

etyloppihacilem commented:
Hi, thank you for your feedback !

I removed all logic from clangd `getSignature` as you suggested, this approach is cleaner. I removed `CK_FunctionQualifier` as well.

At the moment, `IsAddressOfOperand` is needed as `FunctionCanBeCalled` is not able to determine whether a completion is a call or a pointer.

I realized that the _class scope simulation_ was not directly related to this PR anymore, let me know if you want me to remove those modification and open a different PR for clangd/clangd#880.

It can be found at `SemaCodeComplete.cpp:6969` and now uses `Sema::ContextRAII`:
```cpp
std::optional<Sema::ContextRAII> SavedContext;
// When completing a definition, simulate that we are in class scope to access
// private methods.
if (IsInDeclarationContext && Ctx != nullptr)
  SavedContext.emplace(SemaRef, Ctx);
```

etyloppihacilem committed:
Fix of improper use of std::optional

HighCommander4 committed:
Remove CodeCompleteResult::IsAddressOfOperand

HighCommander4 commented on clang/lib/Sema/SemaCodeComplete.cpp:
```diff
@@ -1503,8 +1506,27 @@ void ResultBuilder::AddResult(Result R, DeclContext *CurContext,
-  R.FunctionCanBeCall = canFunctionBeCalled(R.getDeclaration(), BaseExprType);
+  R.DeclaringEntity = IsInDeclarationContext;
+  R.IsAddressOfOperand = IsAddressOfOperand;
+  R.FunctionCanBeCall = canFunctionBeCalled(R.getDeclaration(), BaseExprType) ||
```
In https://github.com/HighCommander4/llvm-project/commit/27df471cb6e0ad83b3d9a9a4a38998686ffb7c70, I'm suggesting two changes to the computation of `FunctionCanBeCall` in two ways.

The first is to add `|| !IsAddressOfOperand`, as a way of avoiding needing to store the `IsAddressOfOperand` flag separately.

I do think that changes deserve a comment such as this:

```c++
// If the user wrote `&` before the function name, assume the
// user is more likely to take the address of the function rather
// than call it and take the address of the result.
```

The second is to remove the `|| IsInDeclarationContext` part which does not make sense conceptually (if the user is writing a declaration, it's definitely not a call). I think the reason for this was so that clangd's `getSignature` doesn't [chop off the parameters](https://searchfox.org/llvm/rev/c56aad6b4f52c9088be5847aa7da33d27df0b338/clang-tools-extra/clangd/CodeCompletionStrings.cpp#294-299) in declaration contexts, which I've addressed in my patch in a different way (passing `IncludeFunctionArguments = FunctionCanBeCall || DeclaringEntity` to `getSignature`).

HighCommander4 commented on clang/lib/Sema/SemaCodeComplete.cpp:
```diff
@@ -1503,8 +1506,27 @@ void ResultBuilder::AddResult(Result R, DeclContext *CurContext,
+
+  // We need to force IsAddressOfOperand when completing a ScopeSpecifier
+  // for non static member function that could be a call.
+  if (!IsInDeclarationContext && !IsAddressOfOperand) { // it may be a call
```
I'm finding that with my changes in https://github.com/HighCommander4/llvm-project/commit/27df471cb6e0ad83b3d9a9a4a38998686ffb7c70, this code block is not necessary to pass any tests. I think it can just be removed.

HighCommander4 commented on clang/lib/Sema/SemaCodeComplete.cpp:
```diff
@@ -3273,18 +3307,18 @@ static std::string GetDefaultValueString(const ParmVarDecl *Param,
+    Preprocessor &PP, const PrintingPolicy &Policy,
+    const FunctionDecl *Function, CodeCompletionBuilder &Result,
+    unsigned Start = 0, bool InOptional = false,
+    bool AsInformativeChunk = false, bool IsInDeclarationContext = false) {
```
With https://github.com/HighCommander4/llvm-project/commit/27df471cb6e0ad83b3d9a9a4a38998686ffb7c70, we are passing `AsInformativeChunk = FunctionCanBeCall || DeclaringEntity` and `IsInDeclarationContext = DeclaringEntity`.

It would probably be clearer to just pass `FunctionCanBeCall` and `DeclaringEntity` as parameters directly and compute their `||` inside this function.

HighCommander4 commented on clang/lib/Parse/ParseExpr.cpp:
```diff
@@ -923,7 +923,7 @@ Parser::ParseCastExpression(CastParseKind ParseKind, bool isAddressOfOperand,
                             tok::l_brace)) {
         // If TryAnnotateTypeOrScopeToken annotates the token, tail recurse.
-        if (TryAnnotateTypeOrScopeToken())
+        if (TryAnnotateTypeOrScopeToken(isAddressOfOperand))
```
Please add a sentence to [this comment](https://searchfox.org/llvm/rev/c56aad6b4f52c9088be5847aa7da33d27df0b338/clang/include/clang/Parse/Parser.h#3848-3851) which explains why `isAddressOfOperand` exists, something like:

```
// In addition, isAddressofOperand is propagated to SemaCodeCompletion
// as a heuristic for function completions (to provide different behavior
// when the user is likely taking the address of a function vs. calling it).
```

HighCommander4 commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -4568,7 +4575,23 @@ class Parser : public CodeCompletionHandler {
       bool IsTypename = false, const IdentifierInfo **LastII = nullptr,
       bool OnlyNamespace = false, bool InUsingDeclaration = false,
-      bool Disambiguation = false);
+      bool Disambiguation = false, bool IsAddressOfOperand = false,
```
Please extend the comment above this function to describe the new parameters.

In that comment, please include a disclaimer that these parameters are used by code completion and may not be set by all callers.

Please do the same for `TryAnnotateTypeOrScopeToken` as well.

(Rationale: these functions have a lot of call sites, and I haven't evaluated in detail for each of them whether e.g. `isAddressOfOperand=true` should potentially be passed in that scenario, nor do I have the level of knowledge of the parser code to do that evaluation. Hence the "may not be set by all callers" disclaimer. This is fine for code completion, but we wouldn't want a future change starting to rely on these flags for some semantic/correctness purpose.)

HighCommander4 reviewed (changes requested):
Ok, I finally had a chance to take a complete pass through the patch. Sorry it took so long.

etyloppihacilem committed:
Cleaning useless code and adding comments.

HighCommander4 committed:
Add comment explaining why we set FunctionCanBeCall=false if AddressOfOperand=true
